src/actions/firebase/firebase.ts

```diff
@@ -0,0 +1,167 @@
+	import * as Hub from "../../hub"
+	import * as admin from 'firebase-admin';
+	import * as app from 'firebase-admin/app';
+
+	export class FirebaseAction extends Hub.Action {
+
+	  name = "firebase"
+	  label = "Firebase"
+	  iconName = "airtable/airtable.png"
+	  description = "Use firebase to send push notifications to mobile."
+	  storageBucket = 'gs://looker-9d4d2.appspot.com'
+	  contentType = "image/jpeg"
+	  notificationOptions = {
+      priority: "high",
+      timeToLive: 60 * 60 * 24,
+      contentAvailable: true,
+      mutableContent: true
+	  }
+	  static firebaseAdmin: app.App
+	  params = []
+	  supportedActionTypes = [Hub.ActionType.Query]
+	  supportedFormats = [Hub.ActionFormat.JsonDetail]
+	  supportedFormattings = [Hub.ActionFormatting.Unformatted]
+	  supportedVisualizationFormattings = [Hub.ActionVisualizationFormatting.Noapply]
+
+    static setFirebaseClient() {
+      if (!FirebaseAction.firebaseAdmin) {
+        FirebaseAction.firebaseAdmin = admin.initializeApp({
+        								credential: admin.credential.cert({projectId: process.env.FIREBASE_PROJECT_ID,
+        																   clientEmail: process.env.FIREBASE_CLIENT_EMAIL,
+        																   privateKey: process.env.FIREBASE_PRIVATE_KEY?.replace(/\\n/g, '\n')
+        																   }),
+        								databaseURL: process.env.FIREBASE_DATABASE,
+        							});
+      }
+    }
+
+	  async execute(request: Hub.ActionRequest) {
+      let response = new Hub.ActionResponse({success: true})
+      try {
+        let data: any = {}
+        if (request.formParams.data) {
+          data = request.formParams.data
+          if (!data.alertId) {
+            throw "Need Valid AlertId."
+          }
+        } else {
+          throw "Need valid notification data."
+        }
+        let imageName = await this.uploadImage(request)
+        await this.verifyAndSendMessage(request.formParams, imageName)
+      } catch(e) {
+          response = new Hub.ActionResponse({success: false, message: e})
+      }
+      return new Hub.ActionResponse(response)
+	  }
+
+    async verifyAndSendMessage(params: Hub.ParamMap, imageName: string): Promise<any> {
+      return new Promise(async (resolve, reject) => {
+        let data: any = params.data
+        if (!params.title) {
+          reject("Needs a valid title.")
+        }
+        if (!params.deviceIds) {
+          resolve()
+        }
+        let notification: any = {title: params.title}
+        let notificationData: any = data
+        notificationData.imageUrl = imageName
+        if (params.timeToLive) {
+          this.notificationOptions.timeToLive = parseFloat(params.timeToLive)
+        }
+        if (params.priority) {
+          this.notificationOptions.priority = params.priority
+        }
+        try {
+          const userObj = JSON.parse(params.deviceIds ?? "[]") as any[]
+          let deviceIdObject = params.deviceIds as unknown as any[]
+          if (deviceIdObject.length == 0) {
```

Comment:
can we put checks like this which will lead to not sending notification before image is uploaded to firebase ?

Reply:
If there's no image it will throw in the earlier upload image method.
This should not be executed after